### ISSUE TITLE
fix segfaults

### ISSUE THREAD

**Issue by sbailey** [desihub/redrock#77]
Even after the MPI refactor, redrock still randomly segfaults when running at scale, even for the pure-MPI version (no multiprocessing).  Debug and fix that.

It is difficult to tell exactly where the problem is occurring for the pure-MPI version because all N>>1 ranks die when any one of them dies, without indicating which one was at fault.

For MPI+multiprocessing, earlier versions would have the multiprocessing spawned processes sometimes randomly immediately die before doing any work (and that problem only occurred on Cray machines, ugh).  Since PR #31 refactored multiprocessing shared memory, we haven't had those crashes, but we've continued to have crashes *after* the multiprocessing processes were finished.  It might still be due to some MPI+multiprocessing incompatibility, or it might be a problem in some other library that is affecting both pure-MPI and MPI+multiprocessing.

Using the wrap-redrock MPI wrapper in two_percent_DESI (which spawns system calls to rrdesi to keep them isolated from each other), I isolated the MPI+multiprocessing crashes to [zfind.py line 227](https://github.com/desihub/redrock/blob/d0a32603c9d0c26c792eca18a17629db83c09ff4/py/redrock/zfind.py#L227):
```
            tzfit = astropy.table.vstack(tzfit)
```
To do:
  * check if the crashes persist when using astropy 2.x (I was testing with astropy 1.3.3)
  * refactor to avoid that vstack (e.g. create the full size table at the start and then fill in the rows)
  * if neither of those work, refactor to avoid astropy tables entirely (we make use of their convenience functions for adding columns, and rearranging things, but with a bit more hassle numpy structured arrays would also work).  Consider doing that anyway.
  * if any of those work to fix the MPI+multiprocessing combination, try it also with the pure-MPI version to see if it also fixes the problem there.

Self-assigning; I want to chase this one.

**Comment by sbailey** [comment]
For the record, the pure-MPI test mysteriously died, but didn't include "segfault" in the message.  Debugging continues...
```
<cori ~> srun --cpu_bind=cores -n 1920 -c 2 desi_pipe_run_mpi --first redshift --last redshift
each proc will wait a few seconds before starting, max is 19.2 sec
INFO:pipe_run.py:60:main: starting at Mon Jan 22 10:41:47 2018
INFO:pipe_run.py:61:main: using raw dir /global/cscratch1/sd/sjbailey/desi/dc17b/spectro/redux/dc17b
INFO:pipe_run.py:62:main: using spectro production dir /global/cscratch1/sd/sjbailey/desi/dc17b/spectro/redux/dc17b
INFO:run.py:499:run_steps: running steps redshift to redshift
INFO:run.py:519:run_steps: starting step redshift at Mon Jan 22 10:51:07 2018
srun: error: nid00054: tasks 1408-1423: Aborted
srun: Terminating job step 9681664.1
slurmstepd: error: *** STEP 9681664.1 ON nid00010 CANCELLED AT 2018-01-22T12:55:04 ***
srun: error: nid00054: tasks 1424-1430,1432,1436-1439: Terminated
srun: error: nid00039: tasks 930,949: Terminated
srun: error: nid00077: tasks 1868,1887: Terminated
srun: error: nid00038: task 900: Terminated
srun: error: nid00051: tasks 1312,1319-1320,1343: Terminated
srun: error: nid00043: tasks 1069,1071,1077,1079,1087: Terminated
srun: error: nid00020: tasks 320,328,331,339,351: Terminated
...
```
This was not due to the job timing out since my interactive queue prompt remained well after that srun ended.

**Comment by sbailey** [comment]
I suspected that it was a memory leak and tried fixing it by changing `astropy.io.fits.open(..., memmap=True)` into `memmap=False` (which has been a source of memory leaks in the past).  I monitored the memory usage while running and it seemed stable for hours, and then "Aborted" again after finishing 3792/4194 pix from dc17b:
```
salloc: Nodes nid00[087-094,113,116-127,137-160,168-182] are ready for job
<cori dc17b> date
Tue Jan 23 10:56:50 PST 2018
<cori dc17b> srun --cpu_bind=cores -n 1920 -c 2 desi_pipe_run_mpi --first redshift --last redshift
each proc will wait a few seconds before starting, max is 19.2 sec
INFO:pipe_run.py:60:main: starting at Tue Jan 23 10:57:27 2018
INFO:pipe_run.py:61:main: using raw dir /global/cscratch1/sd/sjbailey/desi/dc17b/spectro/redux/dc17b
INFO:pipe_run.py:62:main: using spectro production dir /global/cscratch1/sd/sjbailey/desi/dc17b/spectro/redux/dc17b
INFO:run.py:499:run_steps: running steps redshift to redshift
INFO:run.py:519:run_steps: starting step redshift at Tue Jan 23 11:04:48 2018
srun: error: nid00160: tasks 1408-1418,1420,1422-1423: Aborted
srun: Terminating job step 9696774.1
srun: error: nid00160: tasks 1419,1421: Aborted
slurmstepd: error: *** STEP 9696774.1 ON nid00087 CANCELLED AT 2018-01-23T13:08:46 ***
srun: error: nid00160: tasks 1424-1432,1436,1439: Terminated
srun: error: nid00181: tasks 1856,1858-1860,1862-1863,1865-1871,1874-1876,1879,1883,1887: Terminated
...
srun: Force Terminated job step 9696774.1
```
Other notes:
1. for the specfiles that did *not* have zbest output, none of them log files from the pipeline.  Even if one rank was unceremoniously killed, I would have expected partial logfiles to be left over from the other ranks.  I don't know where those logs went.
2. For the logfiles that remained (and had zbest output), none of them had "Terminated", "Aborted", or "memory", or "oom" in them.
3. All of them did end with:
```
--- Logging error ---
Traceback (most recent call last):
  File "/global/common/software/desi/cori/desiconda/20170920-1.2.0-spec/conda/lib/python3.6/logging/__init__.py", line 994, in emit
    stream.write(msg)
ValueError: I/O operation on closed file.
Call stack:
  File "/global/cscratch1/sd/sjbailey/desi/code/desispec/bin/desi_pipe_run_mpi", line 37, in <module>
    pipe_run.main(args, comm=comm)
  File "/global/cscratch1/sd/sjbailey/desi/code/desispec/py/desispec/scripts/pipe_run.py", line 66, in main
    pipe.run_steps(args.first, args.last, spectrographs=args.spectrographs, nightstr=args.nights, comm=comm)
  File "/global/cscratch1/sd/sjbailey/desi/code/desispec/py/desispec/pipeline/run.py", line 521, in run_steps
    grph, ntask, failtask = run_step(step_types[st], grph, opts, comm=comm)
  File "/global/cscratch1/sd/sjbailey/desi/code/desispec/py/desispec/pipeline/run.py", line 265, in run_step
    worker.run(tgraph, tasks[t], options, comm=comm_group)
  File "/global/cscratch1/sd/sjbailey/desi/code/desispec/py/desispec/pipeline/task.py", line 1000, in run
    log.info(" ".join(com))
Message: 'RUN rrdesi_mpi --output /global/cscratch1/sd/sjbailey/desi/dc17b/spectro/redux/dc17b/spectra-64/110/11069/rrdetails_zbest-64-11069.h5 --zbest /global/cscratch1/sd/sjbailey/desi/dc17b/spectro/redux/dc17b/spectra-64/110/11069/zbest-64-11069.fits /global/cscratch1/sd/sjbailey/desi/dc17b/spectro/redux/dc17b/spectra-64/110/11069/spectra-64-11069.fits'
Arguments: ()
================= Process 1 =================
================= Process 2 =================
================= Process 3 =================
================= Process 4 =================
================= Process 5 =================
================= Process 6 =================
================= Process 7 =================
================= Process 8 =================
================= Process 9 =================
================= Process 10 =================
================= Process 11 =================
================= Process 12 =================
================= Process 13 =================
================= Process 14 =================
================= Process 15 =================
```

**Comment by sbailey** [comment]
I had posted a comment about this, but it looks like it got deleted or maybe I posted it to the wrong ticket?  Trying again:

Using desiconda 20170920-1.2.0-spec with astropy 2.0.1 and numpy 1.13.1, the MPI+multiprocessing combination worked for hundreds of spectra files, whereas with desiconda 20170613-1.1.4-spectro with astropy 1.3.3 and numpy 1.11.3, it would fail ~3% of the time.  I have not tried running all 4194 pixels of dc17b using that combination of desiconda/astropy/numpy, however.

**Comment by rainwoodman** [comment]
As far as my memory goes, numpy 1.11.3 had some issues with dtype reference
counting. I also recall since 1.13.x numpy was mostly valgrind clean.

I strongly suspect this may have explained the random crashes you've seen:

https://github.com/numpy/numpy/commit/5500019647bc8138d4eac294016660a6832a2706

**Comment by sbailey** [comment]
And for the record, I have been able to reproduce the "Segmentation fault" that @julienguy reported when using desiconda/20170613-1.1.4-spectro + astropy/1.3.3 + numpy/1.11.3 .  It happened much faster than the "Aborted" message I got when running with newer desiconda/astropy/numpy.

Using desiconda 20180102-1.2.2-spec which has an upgraded numpy (1.13.3) but the same version of astropy (1.3.3), I do *not* get the segmentation fault.  i.e. as @rainwoodman suspected, the bug appears to be in numpy not in astropy.

Summarizing:
* it appears that upgrading numpy fixes the "Segmentation fault" and the MPI+multiprocessing crashes.
* it does not completely fix the pure-MPI case, which still dies with "Aborted" but no informative logfiles that I can find.
* I'm having memory trouble with the largest dc17b pix with MPI+multiprocessing.  I'm confused about this because I previously succeeded in running dc17b redshifts using MPI+multiprocessing...

**Comment by sbailey** [comment]
Thorsten and Rollin at NERSC looked at the "Aborted" jobs and tracked that they were being killed by a signal 15, as if the kill signal had been sent from an external source (!).  They confirmed that it was not an out-of-memory error (which would have been a signal 9).  They suggested adding a python signal handler at the beginning of the program to catch the signal.  That provides a few seconds for the program to dump basic info like where it is in the code and perhaps which rank received the signal before it is actually killed.  What is sending the signal is still just guessing, but that might give us more clues.

**Comment by sbailey** [comment]
Update: the MPI+multiprocessing memory problems were due to a slurm bug that was fixed last night.  Today I was able to run all 2.3M targets in 4193 pixels from dc17b in 3h40m on 60 nodes using the wrap-redrock script from two_percent_DESI and desiconda 20180102-1.2.2-spec with numpy 1.13.3.  There we no failures due to segfaults or aborts, though the wrapper script also has the ability to retry failures such that it also previously worked (with retries on ~3% of the pixels) using numpy 1.11.3.

I'm going to proceed with cleaning up and adding the wrap-redrock script to redrock instead of two_percent_DESI.  That will at least allow us to run redshifts at scale, while continuing to debug the mystery aborts from the pure-MPI version.

**Comment by moustakas** [comment]
How long did it take before the refactor?

**Comment by sbailey** [comment]
The refactor (PR #67) made the multiprocessing version 1.2x faster than before, and the MPI version 1.8x faster on a single test file; I don't have the benchmark numbers for all of dc17b because previous failures required it to be processed in multiple batches.  The main point of this run wasn't raw speed, but rather that this was the first time that we have be able to run this many spectra in a single job without any failures.

**Comment by sbailey** [comment]
Summary: the segfaults were due to a numpy 1.11.3 bug, fixed by 1.13.3.  MPI+multiprocessing works but pure MPI still has mystery "Aborted" crashes, which I spun into a new ticket #100.  Closing this one.

**Comment by rainwoodman** [comment]
Consider running it with ddt to get a stack trace.

On Feb 14, 2018 6:13 AM, "Stephen Bailey" <notifications@github.com> wrote:

Summary: the segfaults were due to a numpy 1.11.3 bug, fixed by 1.13.3.
MPI+multiprocessing works but pure MPI still has mystery "Aborted" crashes,
which I spun into a new ticket #100
<https://github.com/desihub/redrock/issues/100>. Closing this one.

—
You are receiving this because you were mentioned.

Reply to this email directly, view it on GitHub
<https://github.com/desihub/redrock/issues/77#issuecomment-365423152>, or mute
the thread
